### DISCUSSION TITLE
Project Ideas

### DISCUSSION THREAD

**Discussion by jmellorcrummey** [HPCToolkit/hpctoolkit#573] [Ideas]
- Root cause analysis, diagnosis, and optimization of GPU performance bottlenecks
	- Merge GPA analysis capabilities into HPCToolkit and generalize for other GPU architectures. Of particular interest is memory latency diagnosis and optimization.
	- Extend GPA capabilities with support for analysis and optimization of branch divergence, warp synchronization, and tensor core instructions.
	- Extend HPCToolkit's hpcviewer GUI with a new pane for GPU performance diagnosis and optimization suggestions.
- Accelerate support for instrumentation-based GPU instruction-level performance measurement and analysis
	- Turn basic-block level performance measurements on Intel GPUs into instruction-level measurements during post-mortem analysis.
	- Work with new GT-Pin on instrumentation of GPU binaries for instruction-level performance measurement.
- Design and implement concurrent data structures for efficient performance measurement of multithreaded programs
	- Support many-to-one coordination between GPU threads and application threads
	- Identify coordination bottlenecks, design concurrent data structures that address them, and integrate them into HPCToolkit's hpcrun measurement infrastructure
- Measurement, analysis, and diagnosis of memory-related performance bottlenecks
	- identify rate-limiting factors: instruction mix, memory parallelism, etc.
	- characterize performance losses caused by memory latency
	- characterize performance losses caused by limited memory bandwidth (roofline analysis)
		- especially for  GPU codes 
	- extend HPCToolkit's hpcviewer with GUI support for a roofline performance view for  CPU or GPU code
- Support for instruction-level analysis of performance data
	- GUI support for a machine code view based on Capstone IR
	- Add support to Capstone for cracking machine instructions for AMD GPUs
- Accelerate dataflow analysis of instruction-level performance analysis in Dyninst
	- Add support to Dyninst for basic-block level dataflow analysis.

**Comment by wyphan** [comment]
* Initial prototype for a new `hpcreport` tool
  * Sort of like [Intel Application Performance Snapshot][intel-aps] or [Arm Performance Reports][arm-perf]
  * Initially text-only ASCII output, but will extend to a full-blown GUI later
  * Need to record the information in a common format (YAML? JSON? TOML?)
 
* Initial support for Python
  * Waiting for `hpcprof2` merge
  * Extend Jonathon's and Keren's prototype for PyTorch to other frameworks (e.g. TensorFlow, Keras, etc.)
  * Add support for [NVIDIA cuNumeric][cunumeric] (drop-in Legate-based GPU-enabled library to replace NumPy)

[intel-aps]: https://www.intel.com/content/www/us/en/develop/documentation/get-started-with-application-performance-snapshot/top.html
[arm-perf]: https://developer.arm.com/tools-and-software/server-and-hpc/debug-and-profile/arm-forge/arm-performance-reports
[cunumeric]: https://developer.nvidia.com/cunumeric
